### PR TITLE
Update state management to use Universal ID

### PR DESCRIPTION
**Pull request by hopsoft**
## Reduce potential ISP/Gateway privacy concerns related to user tracking etc.

> [!NOTE] 
> There aren't any inherent privacy issues related to TurboBoost Commands (TBC). TBC **does NOT track** user/visitor data for any purpose. The original cookie and header management simply served to facilitate a better user experience (UX).

This PR updates the "state" mechanics to use the URL and hidden form fields rather than cookies and headers for this purpose. 

TurboBoost state management provides a facility for remembering and restoring the state used to render the last representation of a page or component. It also provides the facility to support client side optimistic updates that we can resolve server side. All of this tooling enables developers to provide a great reactive user experience to visitors with minimal code. 

---

After more discovery and some conversations with AI, I've teased out the following areas that are potential spots that may can appear suspect to naive privacy tools.

- Storing larger values in cookies
- Using too many cookies
- Attaching larger values to HTML request headers
- Using too many custom request headers
- Updating or mutating an HTML meta tag too frequently

This PR addresses each of the above items to de-risk privacy hawks from attaching a false positive flag due to any library activity or operations.

### CODE REVIEW

File: app/javascript/turbo.js [hunk 1]
```diff
@@ -12,17 +11,11 @@ addEventListener('turbo:before-fetch-request', event => {
...
     let acceptHeaders = ['text/vnd.turbo-boost.html', fetchOptions.headers['Accept']]
     acceptHeaders = acceptHeaders.filter(entry => entry && entry.trim().length > 0).join(', ')
     fetchOptions.headers['Accept'] = acceptHeaders
-    fetchOptions.headers['TurboBoost-Token'] = meta.token
   }
-
-  // always send state
-  state.payloadChunks.forEach((chunk, i) => {
-    fetchOptions.headers[`TurboBoost-State-${i.toString().padStart(4, '0')}`] = chunk
-  })
```
Comment (hopsoft) [inline app/javascript/turbo.js:25]: Previously we always sent state. We may want to consider bringing this back when we put more attention on state management and optimistic updates.

File: test/dummy/config/importmap.rb [hunk 2]
```diff
@@ -2,19 +2,25 @@
...
 
-pin "@hotwired/stimulus", to: "https://ga.jspm.io/npm:@hotwired/stimulus@3.2.2/dist/stimulus.js"
-pin "@hotwired/turbo", to: "https://ga.jspm.io/npm:@hotwired/turbo@7.3.0/dist/turbo.es2017-esm.js"
-pin "@hotwired/turbo-rails", to: "https://ga.jspm.io/npm:@hotwired/turbo-rails@7.3.0/app/javascript/turbo/index.js"
-pin "@rails/actioncable/src", to: "https://ga.jspm.io/npm:@rails/actioncable@7.1.3/src/index.js"
-pin "@turbo-boost/streams", to: "https://ga.jspm.io/npm:@turbo-boost/streams@0.0.8/app/assets/builds/@turbo-boost/streams.js"
-pin "debounced", to: "https://ga.jspm.io/npm:debounced@0.0.5/src/index.js"
+# TODO: I have no idea why the fuck jspm.io refuses to build the latest versions of @turbo-boost libs
+#       You can generate URLs like the one I'm using below here → https://www.jsdelivr.com/github
+pin "@turbo-boost/streams", to: "https://cdn.jsdelivr.net/gh/hopsoft/turbo_boost-streams@v0.1.6/app/assets/builds/%40turbo-boost/streams.js"
```
Comment (mhenrixon) [inline test/dummy/config/importmap.rb:13]: After merging my [PR](https://github.com/hopsoft/turbo_boost-streams/pull/51), you should be able to use:

```ruby
pin "@turbo-boost/streams", to: "@turbo-boost/streams.js"
```

File: lib/turbo_boost/commands/runner.rb [hunk 3]
```diff
@@ -198,31 +191,36 @@ def handle_command_event(*args)
...
+  # TODO: revisit command token validation
   def new_token
-    @new_token ||= SecureRandom.urlsafe_base64(12)
+    @new_token ||= SecureRandom.alphanumeric(13)
   end
 
+  # TODO: revisit command token validation
   def server_token
-    cookies.encrypted["turbo_boost.token"]
+    nil
```
Comment (mhenrixon) [inline lib/turbo_boost/commands/runner.rb:208]: Is this why elements broke for me perhaps?

**Review by mhenrixon — APPROVED** [review]
As I wrote you in our private conversation, my only problem seems to be with elements that could be expected with such a refactoring. 

Commands seems to be working great for me (I use that individually for some really cool stuff as well).

My PRs should help simplify the setup for testing import maps, but I see nothing obvious in this PR, nor do I have any issues other than with elements. 

It's good to go unless you depend on elements being fixed first.

File: lib/turbo_boost/commands/state.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  def each
+    data.keys.each do |key|
+      yield key, self[key]
+    end
+  end
+
+  # TODO: implement state resolution
+  def resolve(client_state)
+    # return unless self.class.resolver
```
[the remark this answers is not in the document]
Comment (hopsoft) [reply lib/turbo_boost/commands/state.rb:35]: I'm happy to pair more and walk you though my thinking on this; however, I just wanted to get the stub in place so we can revisit easily and make future improvements.

File: test/state_test.rb [hunk 5]
```diff
@@ -3,99 +3,69 @@
 require "test_helper"
 
 class StateTest < ActiveSupport::TestCase
-  test "empty state" do
-    state = TurboBoost::State.new
-    assert_equal 0, state.size
-    assert_equal 3, state.payload.bytesize
-    assert_equal 16, state.ordinal_payload.bytesize
+  test "new with defaults" do
```
[the remark this answers is not in the document]
Comment (hopsoft) [reply test/state_test.rb:6]: True in this case... unless we ever want to establish a default baseline state.

File: lib/turbo_boost/commands/state.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def cache_key
+    "TurboBoost::Commands::State/#{Digest::SHA2.base64digest(to_json)}"
+  end
+
+  def read(...)
+    value = store.read(...)
+    value = parent&.read(...) if value.nil?
+    value
```
[the remark this answers is not in the document]
Comment (hopsoft) [reply lib/turbo_boost/commands/state.rb:50]: Actually I had that to begin with but there's a situation where you may assign false that we want to honor; hence, the explicit nil check vs a falsey one.

File: lib/turbo_boost/commands/runner.rb [hunk 7]
```diff
@@ -258,11 +256,16 @@ def append_streams_to_response_body
     command_instance.turbo_streams.each { |stream| append_to_response_body stream }
   end
 
-  def append_meta_tag_to_response_body
-    cookies.encrypted["turbo_boost.token"] = {value: new_token, path: "/"}
-    append_to_response_body turbo_stream.invoke("morph", args: [meta_tag], selector: "#turbo-boost")
+  def append_command_token_to_response_body
```
[the remark this answers is not in the document]
Comment (hopsoft) [reply lib/turbo_boost/commands/runner.rb:259]: Perhaps we can follow up with more clear semantic naming for some of these.

File: test/state_test.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (julianrubisch) [reply test/state_test.rb:6]: yes, we could define an adapter interface, but as it stands this test would brake if we ever swap the internals

**Review by BuddyLReno — APPROVED** [review]
I don't have enough context to offer any meaningful feedback but I didn't see any obvious logic errors or typos 👍

**Pull request merged**